scripts/build_containers.sh

```diff
@@ -102,6 +102,56 @@ function create_production_containers {
 
 }
 
+function create_release_containers {
+  BRANCH=$1
+
+  # Pull images if available for caching
+  docker pull "${CONTAINER_REPO}"/"${CONTAINER_APP}":builder ||:
+  docker pull "${CONTAINER_REPO}"/"${CONTAINER_APP}":production ||:
+  docker pull "${CONTAINER_REPO}"/"${CONTAINER_APP}":testing ||:
+  docker pull "${CONTAINER_REPO}"/"${CONTAINER_APP}":development ||:
+
+  # Build the builder image
+  docker build --target builder \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":builder \
+               --tag "${CONTAINER_REPO}"/"${CONTAINER_APP}":builder .
+
+  # Build the production image
+  docker build --target production \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":builder \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":production \
+               --tag "${CONTAINER_REPO}"/"${CONTAINER_APP}":${BUILDKITE_COMMIT:0:7} \
+               --tag "${CONTAINER_REPO}"/"${CONTAINER_APP}":${BRANCH} .
+
+  # Build the testing image
+  docker build --target testing \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":builder \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":production \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":testing \
+               --tag "${CONTAINER_REPO}"/"${CONTAINER_APP}":testing-${BRANCH} .
+
+  # Build the development image
+  docker build --target development \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":builder \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":production \
+               --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":testing \
+               --tag "${CONTAINER_REPO}"/"${CONTAINER_APP}":development-${BRANCH} .
+
+  # If the env var for the git tag doesn't exist or is an empty string, then we
+  # won't build a container image for a cut release.
+  if [ -v BUILDKITE_TAG || ! -z "${BUILDKITE_TAG}" ]; then
+    docker build --target production \
+                 --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":builder \
+                 --cache-from="${CONTAINER_REPO}"/"${CONTAINER_APP}":production \
+                 --tag "${CONTAINER_REPO}"/"${CONTAINER_APP}":${BUILDKITE_TAG} .
+  fi
```

Comment:
potentially dumb questions, what is a  "cut" release, and how is this section different from L120-L124?

Reply:
@maestromac Not dumb questions at all 🙂 

> what is a "cut" release

It's just a released version. Maybe I should've written "tagged release" there, but it all means the same thing. If you think about a branch as a stream of commits, tagging a release effectively cuts off that stream for a specific version, so it's often referred to as "cutting" a release.

> how is this section different from L120-L124?

This block builds a container for the release tag. So, for example, if we were to cut a release today, we would tag it as `stable.20210726.0`. The `stable` branch (and the `stable` container tag) will get updated when the branch receives new commits, but the release tag stays because it's associated with a specific commit and is immutable.

Since Buildkite only populates the `BUILDKITE_TAG` if we're pushing a tag, pushing untagged git commits to `stable` will build a new `stable` container, but will not invoke this block because there is no git tag. If we're pushing tagged commits (as [`scripts/release` does](https://github.com/forem/forem/pull/13750/files/691dc32868ce33efce55fe19db220acf11d87e5a#diff-738cff26e36d290aec109d76d56425fdf8ba8ac9c4aae6b8ecc65b79d484e6ecR88)), it will build both the `stable` and `stable.YYYYMMDD.0` container tags.

Reply:
Ah that makes sense now. Thank you!